blog/2025-11-03-conda-pip-docker-nix.mdx

```diff
@@ -0,0 +1,144 @@
+
+---
+
+### Shared techniques with Nix: prefix-relocatability and patching
+
+Both **conda packages** and **Nix** instrument builds so binaries are **prefix-aware** and relocatable:
+
+- **[RPATH](https://en.wikipedia.org/wiki/Rpath)/[RUNPATH](https://en.wikipedia.org/wiki/Rpath) patching** so [ELF](https://en.wikipedia.org/wiki/Executable_and_Linkable_Format) binaries load from `$PREFIX/lib`
+- **[Shebang](https://en.wikipedia.org/wiki/Shebang_(Unix)) rewriting** so scripts execute with the environment's interpreters
+- **[Mach-O](https://en.wikipedia.org/wiki/Mach-O) (macOS) fix-ups** (`install_name_tool`) for `LC_LOAD_DYLIB`/`LC_ID_DYLIB`
+- **ELF surgery via [`patchelf`](https://github.com/NixOS/patchelf)** *(the tool originated in the Nix ecosystem and is widely used by `conda-build`)*
+
+The result is the same idea in both worlds: build products that **bind to their own prefix**, making environments movable and reproducible. Unlike conda’s goal of relocatable prefixes, Nix intentionally uses content-hashed, fixed store paths to guarantee purity; relocation is not a design objective there.
+
+---
+
+### Why not shipping `glibc` keeps envs small (and portable)
+
+**Conda's design philosophy—not shipping `glibc`** (or the OS C runtime)—has big consequences:
+
+1. **No base OS duplication.** Containers/Nix closures ship `libc`, `ld-linux`, locales/NSS, and friends; that's tens to hundreds of MB before your app starts. **Conda packages** reuse the host `libc`, so environments don't carry that weight.
+
+2. **Fewer cascading deps.** Pulling in `glibc` drags a constellation of tightly-coupled runtime pieces (dynamic loader, NSS, timezone/locale data). Leaving that to the host avoids inflating **every** environment.
+
+3. **Security & updates.** Host OS updates to `glibc` apply automatically across conda envs. **The conda package managers** (conda, mamba, pixi) enforce compatibility using **virtual packages** (e.g., `__glibc>=2.17`) and by **building against the oldest supported `glibc`** for forward compatibility.
+
+4. **Still safe and predictable.** The solver includes the host fact (`__glibc`, `__osx`, `__win`) so you cannot accidentally solve to an incompatible set.
+
+**Net effect:** environments that are typically **hundreds of MB smaller** than container images with equivalent stacks, and far faster to create, copy, and cache—without giving up ABI coherence.
+
+---
+
+## Conda vs. Pip, Docker, and Nix
+
+Let’s place them side by side:
+
+| Aspect | Pip/npm | Conda | Docker | Nix |
+|--------|---------|-------|--------|-----|
+| **Scope** | Language-specific libraries | Multi-language user-space distribution | Full OS filesystem image | Whole-system functional distribution |
+| **Unit of publish** | Wheel/sdist (library) | Distro package (build + run metadata) | Layered image | Derivation output (hashed closure) |
```

Comment:
`Unit of publish` sounds strange to me as a non-native speaker, how about `Publishable unit` or `Published unit`?